packages/server/src/api/controllers/query/index.ts

```diff
@@ -142,6 +142,78 @@ export async function preview(ctx: UserCtx) {
 
   const authConfigCtx: any = getAuthConfig(ctx)
 
+  function getFieldMetadata(
+    field: any,
+    key: string,
+    nestedSchemaFields: {
+      [key: string]: Record<string, string | QuerySchema>
+    }
+  ): QuerySchema {
+    const makeQuerySchema = (
+      type: FieldType,
+      name: string,
+      subtype?: string
+    ): QuerySchema => ({
+      type,
+      name,
+      subtype,
+    })
+    let type = typeof field,
+      fieldMetadata = makeQuerySchema(FieldType.STRING, key)
+    if (field)
+      switch (type) {
+        case "boolean":
+          fieldMetadata = makeQuerySchema(FieldType.BOOLEAN, key)
+          break
+        case "object":
+          if (field instanceof Date) {
+            fieldMetadata = makeQuerySchema(FieldType.DATETIME, key)
+          } else if (Array.isArray(field)) {
+            const firstFiveItems = field.slice(0, 5)
+            if (firstFiveItems.some(item => JsonUtils.hasSchema(item))) {
+              fieldMetadata = makeQuerySchema(
+                FieldType.JSON,
+                key,
+                JsonFieldSubType.ARRAY
+              )
+            } else {
+              fieldMetadata = makeQuerySchema(FieldType.ARRAY, key)
+            }
+            nestedSchemaFields[key] = buildSchemaFromArray(
+              firstFiveItems,
+              nestedSchemaFields
+            )
+          } else {
+            fieldMetadata = makeQuerySchema(FieldType.JSON, key)
+          }
+          break
+        case "number":
+          fieldMetadata = makeQuerySchema(FieldType.NUMBER, key)
+          break
+      }
+    return fieldMetadata
+  }
+
+  function buildSchemaFromArray(
+    firstFiveItems: any[],
```

Comment:
nit: I wouldn't call this `firstFiveItems`, as there's no such limitation inside this function. `items` is probably sufficiently general.